### PR TITLE
Merge remote-tracking branch 'upstream/stable' into merge_stable

### PR DESCRIPTION
**Pull request by MartinNowak**

### CODE REVIEW

**Comment by andralex** [comment]
@MartinNowak So... what do we do here?

**Comment by MartinNowak** [comment]
> @MartinNowak So... what do we do here?

Seems like the added [AliasDeclaration.isOverloadable] (https://github.com/dlang/dmd/pull/6125/commits/22dc48171eaa3ef43dbec3e6bab6ea7fcb839a37#diff-7eb4e0187665d2ea4b178598a8a47a21R875) doesn't work w/ #5690, b/c selectively imported symbols are now resolved lazily (on master). Need to find a proper solution for that.

Changing isOverloadable to return true while the alias is still unresolved seems problematic (even though that's the internal logic of AliasDeclaration).
As this happens in the visibility check (on usage) we might be able to trigger semantic of the AliasDeclaration.

Hard to believe that we put even more complexity into this horrible AliasDeclaration mess.

**Comment by MartinNowak** [comment]
Added a commit to resolve the lazy aliases in isOverloadable.

**Comment by MartinNowak** [comment]
> Added a commit to resolve the lazy aliases in isOverloadable.

Didn't work of course due to some order constraints, now marked unresolved aliases as overloadable and handled them in the overload visibility check.

File: src/access.d [hunk 1]
```diff
@@ -540,3 +521,83 @@ extern (C++) bool symbolIsVisible(Scope *sc, Dsymbol s)
...
+        if (auto fd = s.isFuncDeclaration())
+            next = fd.overnext;
+        // template temp(T) {}
+        // private template temp(T:int) {}
+        else if (auto td = s.isTemplateDeclaration())
+            next = td.overnext;
+        // alias common = mod1.func1;
+        // alias common = mod2.func2;
+        else if (auto fa = s.isFuncAliasDeclaration())
+            next = fa.overnext;
```
Comment (andralex) [inline src/access.d]: could you please get this line covered?

File: src/access.d [hunk 2]
```diff
@@ -540,3 +521,83 @@ extern (C++) bool symbolIsVisible(Scope *sc, Dsymbol s)
...
+                    assert(ad.overnext is null, "Unresolved overload of alias");
+                    break;
+                }
+            }
+            // handled by ddmd.func.overloadApply for unknown reason
+            assert(next !is ad); // should not alias itself
+            assert(next !is fstart); // should not alias the overload list itself
+        }
+        else
+            break;
```
Comment (andralex) [inline src/access.d]: This line should be covered, too.

File: src/declaration.d [hunk 3]
```diff
@@ -921,6 +965,11 @@ extern (C++) final class OverDeclaration : Declaration
...
     }
 
+    override bool isOverloadable()
+    {
+        return true;
+    }
+
     Dsymbol isUnique()
     {
         if (!hasOverloads)
```
Comment (andralex) [inline src/declaration.d]: etc

**Review by andralex — COMMENTED** [review]
Please let me know whether you'd want to cover the added code with this PR or in a future pass.

**Comment by MartinNowak** [comment]
> Please let me know whether you'd want to cover the added code with this PR or in a future pass.

In a future PR @andralex, this just merges back the stable branch and shouldn't contain additional code. With the next release we should also get coverage tests on stable (would've catched missing test cases earlier).

File: src/declaration.d (hunk 3 above)
Comment (MartinNowak) [inline src/declaration.d]: This one is not part of the diff.

File: src/access.d (hunk 1 above)
Comment (MartinNowak) [inline src/access.d]: Yes, would be good.

**Pull request merged**